# Forecasting Medicaid Prescription Drug Demand
*Capstone project for BrainStation Data Science diploma*

Author - Albert King<br>
Last Modified - 11 April 2023<br>
Contact - awkchemistry@gmail.com<br>

## Notebook 4

Drug shortages directly contribute to preventable deaths, drug stockpiling and rationing by hospitals and governments, and inflated drug costs. In effort to characterize the situation of current drug shortages, this project looks at reported US Medicaid drug claims by states for years 1991-2022, the most recent data currently available. In combination with the FDA's current list of shortages and anticipated shortages, time series analysis is used to forecast future demand.

This project is divided into 4 Jupyter notebooks:

- Notebook 1 imports Medicaid data into a MySQL database, explores and cleans prescription data, and prepares a list of drugs common to the Medicaid data and current FDA shortages.
- Notebook 2 explores and models the prescription data aggregated to total prescriptions per year nationwide. Fitting is accomplished through ARIMA and Facebook Prophet models.
- Notebook 3 models the data using an ARIMA multiple-point output convolutional neural network (CNN). It is separate due incompatabilities between the environments used for Notebooks 1, 2, and 4 with TensorFlow/Keras.
- Notebook 4 ingests, cleans, models, and forecasts prescription demand for individual drugs en masse.

Medicaid datasets were downloaded locally as CSV files from https://data.medicaid.gov/datasets?theme%5B0%5D=State+Drug+Utilization for all years available. The CSV files were brought into MySQL initially using command line as MySQLWorkbench was not able to handle the size of the files. This method is recreated below so as to aggregate methods into Jupyter notebooks, though the initial command line code is included in an appendix. Current drug shortages were downloaded from https://www.accessdata.fda.gov/scripts/drugshortages/default.cfm as CSV files.

Raw data are available at https://1drv.ms/f/s!Ah-g64c_vUtNoCNJ86SVWQk5qyoh?e=GNFof6.

## Bulk Fitting and Forecasting

This notebook digests prescription data en masse. Generally, several functions are written which serve to collect data for each specified drug; clean the data by removing extremely high values and forward filling any empty values; fitting the data to optimize ARIMA parameters; averaging the top 5 fits; and forecasting (unsupervised) future prescriptions for each drug.

This is the final notebook in this project.

## Table of Contents

- [Imports](#imports)
- [Functions](#functions)
- [Fitting the Data](#fitting-the-data)
- [Forecasting](#forecasting)

## Imports

The relevant libraries are called below. Certain values needed throughout the notebook are instantiated.

In [19]:
# Imports and global settings

# regular imports
import pandas as pd
import numpy as np
import seaborn as sb
import matplotlib.pyplot as plt
import plotly.graph_objects as go

from statsmodels.tsa.statespace.sarimax import SARIMAX

from pandas.tseries.offsets import QuarterEnd

# import sqlalchemy to interact with MySQL
import sqlalchemy as db

# working with MySQL generates many warnings - they're hidden here but may be needed on initial setup!
# WARNING WARNING WARNING
import warnings
warnings.filterwarnings("ignore")
# WARNING WARNING WARNING

# table names are listed in a dict with their corresponding years for convenience converting between the two
tables = {'d91':'1991','d92':'1992','d93':'1993','d94':'1994','d95':'1995','d96':'1996','d97':'1997',
          'd98':'1998','d99':'1999','d00':'2000','d01':'2001','d02':'2002','d03':'2003','d04':'2004',
          'd05':'2005','d06':'2006','d07':'2007','d08':'2008','d09':'2009','d10':'2010','d11':'2011',
          'd12':'2012','d13':'2013','d14':'2014','d15':'2015','d16':'2016','d17':'2017','d18':'2018',
          'd19':'2019','d20':'2020','d21':'2021','d22':'2022'}

# drugs = ['acetamin']
# drugs = ['abacavir']

# initializing parameters to access MySQL
my_username = "root"
my_password = "rootroot"
schema_name = "medicaid" #note that schema has already been created in MySQL dashboard directly

# defining engine for connection
engine = db.create_engine(f"mysql+pymysql://{my_username}:{my_password}@localhost/{schema_name}", echo=False)

# defining connection
conn = engine.connect()

The list of top drugs is imported.

In [20]:
# import drugs in Medicaid and shortage dataset
drugs = pd.read_csv('data/matched.csv')

# convert pd series to list
drugs = list(drugs['0'])

# check 80 present
len(drugs)

336

Notebook 1 counted 338 drugs similar between the Medicaid and shortage lists. Due to using a dictionary in assembling the list of matching drugs, it is inferred that there are 336 drugs present because duplicate keys cannot be repeated, suggesting apparent duplicates present in the shortage list. As data have been called, I'll now prepare the functions to manage the data.

## Functions

I first create functions to evaluate fits, name Mean Absolute Percentage Error and Symmetric Mean Absolute Percentage Error, respectively MAPE and SMAPE. The SMAPE is more robust to zeros in data, whether in the training/test or predicted data. The MAPE function is included as it may be useful to compare specific predictions.

In [21]:
def mean_absolute_percentage_error(true_values, predicted_values):
    """
    Input is series of known and then predicted values from time series.
    
    Calculate the mean absolute percentage error.
    
    Find the prediction error and divide by the true value, then average.
    """
    
    error = true_values - predicted_values
    
    absolute_percentage_error = np.abs(error/true_values)
    
    mape = absolute_percentage_error.mean() * 100
    
    return mape

def symmetric_mean_absolute_percentage_error(true_values, predicted_values):
    """
    Input is series of known and then predicted values from time series.

    Calculation is more robust to zeros in data, such as years with zero prescriptions
    or projections of zero.

    Calculate the symmetric mean absolute percentage error.
    
    Find the prediction error and divide by the true+predicted value, then average.
    """

    # calculate numerator - difference in values
    numerator = abs(predicted_values-true_values)
    
    # calculate denominator, sum of absolute true/predicted
    denominator = abs(true_values)+abs(predicted_values)
    
    division = numerator/denominator
    
    summed_div = division.sum()
    
    # average
    smape = (100/(len(true_values)))*summed_div
    
    return smape

Next, a function is created to search MySQL for drug specific data. The drug is matched using a regular expression, prescriptions are summed across each quarter for each year, and the data are converted to a dataframe.

In [22]:
# make summary sql table

def drug_search(drug):
        """
        Input is name of drug to be matched in Medicaid dataset as string.

        Drug then matched as regex by iterating through years of available MySQL data,
        grouping by quarter and year.

        Output is a pandas dataframe with datetime index and number of reimbursed prescriptions in a single column.
        """
    
        # empty dictionary for data to be held
        drug_dict = {}
    
        drug = drug
    
    # iterate through tables/years, each pulling out prescriptions summed for the drug by quarter
        for table_name in tables.keys():

                # create query for MySQL
                query = f"SELECT SUM(number_of_prescriptions) AS 'Prescriptions', quarter FROM {table_name} WHERE product_name REGEXP '^{drug}' GROUP BY quarter ORDER BY quarter ASC;"

                # run query, result returns as dataframe
                df = pd.read_sql(query, conn)

                # generating the date by quarter for each entry
                year = tables[table_name]

                # loops to convert quarter into month/day format
                # assumes regular quarters
                for i in df.index:
                        qtr = df['quarter'][i]

                        if qtr == 1:
                                month_day = '03/31'
                        elif qtr == 2:
                                month_day = '06/30'
                        elif qtr == 3:
                                month_day = '09/30'
                        elif qtr == 4:
                                month_day = '12/31'

                        # fills dictionary of date/prescriptions
                        drug_dict[f"{month_day}/{year}"] = df.iloc[i,0]

        # total date/prescriptions aggregated in dataframe
        drug_df = pd.DataFrame.from_dict(drug_dict, orient='index', columns=['Prescriptions'])

        # ensure that index is datetime type
        drug_df.index = pd.to_datetime(drug_df.index)
    
        # return dataframe
        return drug_df

The next step will be to clean up the data. As some contain erratic values, values greater than 5 times the median will be removed. All blanks will be interpolated via the fill forward method. From my experience in the first notebook, I have chosen not to use a Tukey's test for outliers because I observed it to be very sensitive operating on related data, and I am concerned that it would result in too much data being interpolated. A future iteration will employ more rigorous statistical methods to assess outliers and replace as appropriate.

In [23]:
def auto_clean(drug_df):
    """
    Input is dataframe with datetime index and number prescriptions as values.

    Cleans data by checking for values > 5* the dataset median;
    if found, values are changed to none.

    All blank/None values are filld by fillforward method.
    """

    drug_df = drug_df
    
    # find the median of the data
    median = drug_df['Prescriptions'].median()
    
    # iterate through each row and check if value should be replaced
    for i in drug_df.index:

        # if value > 5* median, value removed
        if drug_df['Prescriptions'][i] > median*5:
            drug_df['Prescriptions'][i] = None
    
    # empty values interpolated by fillforward
    drug_df.interpolate(method='ffill', inplace=True)
    
    # returns cleaned dataframe
    return drug_df

I want to fairly seamlessly convert data between functions, so I created a class to hold data from the fits so that it may be easily accessed later.

In [24]:
class Parameter:
    """
    Defines a class to contain parameter_fitting function data.
    Inputs are start, end, and pdqs, which must be a Pandas dataframe.
    Methods are .start_stop and .pdq_values,
    which respectively contain a list of start and stop dates
    and a pandas dataframe containing the top 5 optimized
    ARIMA p, d, and q values based on minimized SMAPE.
    """

    def __init__(self, start, end, pdqs: pd.DataFrame):
        self.start_stop = (start, end)
        self.pdq_values = pdqs

The next function iterates through ARIMA p, d, and q parameters to minimize SMAPE. Each set of parameters is fit to the train data, ued to predict the full dataset, and compared to the test dataset. The top 5 ARIMA parameters are returned for use in forecasting future values.

In [49]:
def parameter_fitting(drug_df,end_dataset,start_dataset='1991/03/31'):
    """
    A function to fit the data, test p, d, q values, choose the top 5 based
    on minimized SMAPE, and returns a Parameters class variable.

    Inputs are cleaned drug dataframe, requested end date, and requested
    start date. Start date defaults to earliest data, 1991/03/31.
    
    Possible p, d, and q values are iterated through, the SMAPE calculated
    for all options, and the top five lowest SMAPE values chosen to return.
    """

    drug_df = drug_df

    # cutoff between train and test, inclusive to train
    cutoff = pd.to_datetime(end_dataset)

    # if the dataset needs shortened i.e. due to missing early time data this can be implemented
    start = pd.to_datetime(start_dataset)
    
    # instantiate empty dictionaries for train/test ARIMA parameters and SMAPEs
    train_dict = {}
    test_dict = {}
    
    # instantiating a counter to order data
    run_num = 0

    # iterates through p, d, and q values sequentially, from 0-10
    for p in range(0,6):
        for d in range(0,6):
            for q in range(0,6):

                # the process can be quite slow so there's an updating print statement
                print(f"Running series p {p}, d {d}, q {q}",end="\r")
                
                # some ARIMA combinations are invalid or give infinite values; a try statement
                # is used so that the iterations can continue even if such an error arises.

                    
                # copy dataframe so it's not accidentally altered
                predict_df = drug_df.copy(deep=True)

                # ensure datetime index
                predict_df.index = pd.to_datetime(predict_df.index)
                predict_df.dropna(axis=0, inplace = True)

                # train and test are defined based on cutoff date
                train = predict_df.loc[predict_df.index <= cutoff].dropna()
                test = predict_df.loc[predict_df.index > cutoff]

                # SARIMAX model is trained and fit with iteration's parameters
                model = SARIMAX(train, order=(p, d, q), trend="c")
                model_fit = model.fit(disp=0)

                # predictions are made for the full dataset
                predictions = model_fit.predict(start=0, end=len(train)+len(test)-1)

                # keeping the datetime index - unsure why it was being lost
                predicted_vals = list(predictions)

                # predicted data as column with datetime index
                predictions = pd.Series(predicted_vals, index=list(predict_df.index))

                # train and test SMAPE are calculated
                train_smape = symmetric_mean_absolute_percentage_error(pd.Series(list(train[train.columns[0]])), pd.Series(list(predictions[train.index])))
                test_smape = symmetric_mean_absolute_percentage_error(pd.Series(list(test[test.columns[0]])), pd.Series(list(predictions[test.index])))

                # train and test SMAPE are put in dictionaries with corresponding iteration parameters
                # NOTE THAT 'val' CONTAINS SMAPE
                train_dict[run_num] = {'p':p,'d':d,'q':q,'val':train_smape}
                test_dict[run_num] = {'p':p,'d':d,'q':q,'val':test_smape}
                
                # run is iterated
                run_num+=1
                


    # dataframes are instantiated for train and test data
    train_vals_df = (pd.DataFrame(train_dict).T)
    test_vals_df = (pd.DataFrame(test_dict).T)

    # an empty dict made to hold top parameters
    combined_dict = {}

    # iterates through train dataset to find lowest SMAPE values
    for i in train_vals_df.index: # make a list of places where they overlap then plot
        
        # strips out only positive and <100% SMAPE values for train and test
        # NOTE I don't think the 0 < clause is needed? - should review (or just use absolute values)
        if 0 < train_vals_df.loc[i,'val'] < 100 and 0 < test_vals_df.loc[i,'val'] < 100:
            
            # instances where SMAPE is optimized for both train and test are combined in a dataframe.
            combined_dict[i] = {'p':train_vals_df.loc[i,'p'],
                                'd':train_vals_df.loc[i,'d'],
                                'q':train_vals_df.loc[i,'q'],
                                # note that average SMAPE between train and test is stored
                                'val':(train_vals_df.loc[i,'val']+test_vals_df.loc[i,'val'])/2}
    
    try:
        # optimized parameter dict to dataframe
        combined_df = pd.DataFrame(combined_dict).T
        
        # datafrae ordered by averaged SMAPE
        combined_df.sort_values(by='val',ascending=True)
        
        # the top 5 parameters are retained
        top_combined = combined_df.sort_values('val')[0:6].reset_index(drop=True)
        
        # Parameter class file with data created to return
        parameters = Parameter(start,cutoff,top_combined)
    except:
        pass
    
    return parameters

The top 5 parameters are used to average a single fit. In order to convert the averaged data for use in forecasting, the Final_Result class is created.

In [26]:
class Final_Result:
    """
    Class created to hold data from averaged_model function. Contains train and test data,
    train and test SMAPE, the averaged fit, and fit parameters from the top 5 ARIMA parameter fits.
    """
    def __init__(self, train, test, train_eval, test_eval, fit_mean, fit_parameters):
        self.train = train
        self.test = test
        self.mean = fit_mean
        self.fit_parameters = fit_parameters
        self.train_eval = train_eval
        self.test_eval = test_eval

The top 5 parameter sets are used to fit and determine an average, optimized fit. The model is ultimately used for forecasting.

In [27]:
def averaged_model(drug_df, parameters: Parameter):
    """
    Function to average the fit of the top 5 ARIMA parameters, return fit/prediction data.
    Input is cleaned dataframe and parameters which must be Parameter class. Return is Final_Result class.
    """
    
    # instantiate empty dicts for fit and parameters
    avgd_fit_dict = {}
    fit_params = {}
    
    # get p, d, q parameters from Parameters file
    top_combined = parameters.pdq_values.copy(deep=True)
    
    # get cutoff used previously
    cutoff = parameters.start_stop[1]
    
    # copy the cleaned drug dataframe so it's not accidentally modified
    predict_df = drug_df.copy(deep=True)
    
    # ensure datetime index
    predict_df.index = pd.to_datetime(predict_df.index)
    predict_df.dropna(axis=0, inplace = True)

    # define train and test sets based off of cutoff date
    train = predict_df.loc[predict_df.index <= cutoff].dropna()
    test = predict_df.loc[predict_df.index > cutoff]

    # iterate through top 5 parameter sets 
    for ind in top_combined.index:

        # grabs the ARIMA parameters
        p, d, q = int(top_combined.loc[ind,'p']), int(top_combined.loc[ind,'d']), int(top_combined.loc[ind,'q'])

        # instantiate model and fit on train
        model = SARIMAX(train, order=(p, d, q), trend="c")
        model_fit = model.fit(disp=0)

        # predict for full dataset
        predictions = model_fit.predict(start=0, end=len(train)+len(test)-1)

        # retain datetime index
        predicted_vals = list(predictions)
        predictions = pd.Series(predicted_vals, index=list(predict_df.index))

        # store model_fit and predicted fit in dicts
        fit_params[f'p{p}, d{d}, q{q}'] = model_fit
        avgd_fit_dict[f'p{p}, d{d}, q{q}'] = predictions

    # convert predicted fits to dataframe
    avgd_fit_df = pd.DataFrame(avgd_fit_dict, index = list(predictions.index))
    
    # create a mean of the top 5 fits
    avgd_fit_df['mean'] = avgd_fit_df.mean(axis=1)

    # find SMAPE for train/test using averaged fit
    train_smape = mean_absolute_percentage_error(pd.Series(list(avgd_fit_df['mean'][0:len(predictions[train.index])])), pd.Series(list(predictions[train.index])))
    test_smape = mean_absolute_percentage_error(pd.Series(list(avgd_fit_df['mean'][len(predictions[train.index]):])), pd.Series(list(predictions[test.index])))

    # create Final_Result variable containing train, test, MAPEs, mean fit, and fit parameters
    result = Final_Result(train, test, train_smape, test_smape, avgd_fit_df['mean'], fit_params)

    return result

## Fitting the Data

After all functions have been created, it is time to fit data. Each drug is iterated through each of the functions to produce an average fit.

In order to keep the project time manageable, drug data was collected in a CSV, so that fitting could be performed separately from data collection.

In [445]:
# creating dataset

# empty df
matched_data = pd.DataFrame()

# iterates through each matched drug
for drug in drugs:

    # update note
    print(f"Processing {drug} ",end="\r")
    
    # sends drug name to MySQL query, returns dataframe
    drug_df = drug_search(drug)

    # dataframe is cleaned
    drug_df_clean = auto_clean(drug_df)
    
    # drug name given to column for tracking
    drug_df_clean.rename(mapper = {'Prescriptions':f'{drug}'}, axis = 1, inplace = True)
    
    # data to a master df
    matched_data = pd.concat([matched_data, drug_df_clean], axis = 1)

# export to csv
matched_data.to_csv('data/matched_data.csv')

Because the searching process can be so long, the analysis is run separately. Below, only drugs with full columns of data are run, and their aggregate fits forwarded for forecasting.

In [50]:
# dict to hold aggregated data
runs = {}

# import the cleaned data
df = pd.read_csv('/Users/albert/Documents/Capstone/data/matched_data.csv', index_col=0)

# iterate through each drug in the dataset
for column in list(df.columns):
    
    # update progress
    print(f"Currently processing {column}.", "\n", end = "\r")

    # if the data does not have early time data it is passed because it will cause issues with fitting
    if df[column][df.index[0]:df.index[3]].isna().sum() > 0:

        pass
    
    else:

        try:
            # put data in a df
            drug_df_clean = pd.DataFrame(df[column])
            
            # run through parameter fitting function
            parameters = parameter_fitting(drug_df_clean,'2020-12-31', min(drug_df_clean.index))
            
            # optimized model returned
            result = averaged_model(drug_df_clean, parameters)
            
            # aggregated to master dict
            runs[column] = [drug_df_clean, result]
        
        except:
            pass

If desired, a specific drug can be pulled out and plotted to view the fit compared to the data.

In [ ]:
# we can view the data if we want

# prevets error when full dict of drugs present
drug = drugs[0]

# extracts train, test, and mean/averaged fit
train = runs[drug[0]][1].train
test = runs[drug[0]][1].test
mean = runs[drug[0]][1].mean

# plots example data
fig = go.Figure()
fig.add_trace(go.Scatter(x=list(train.index), y = train['Prescriptions'], name = 'Train'))
fig.add_trace(go.Scatter(x=list(test.index), y = test['Prescriptions'], name = 'Test'))
fig.add_trace(go.Scatter(x=list(mean.index), y = mean.values, name = 'Averaged Fit'))
fig.update_layout(title=f"{drug} Prescriptions Annually<Br>Averaged Fit")
fig.update_xaxes(rangeslider_visible=True)
fig.update_layout(xaxis_title = 'Year', yaxis_title='Prescriptions')
fig.show()

## Forecasting

The assembled data can now be used to produce forecasts from the fit data collected above.

In [51]:
# forecasting

# number of steps to forecast - because data are in quarter, 4 = 1 year from cutoff
steps = 4

# instantiate empty dicts to hold data
forecasts = {}
predictions = {}

# iterate through each drug
for run in runs.keys():

    # total length of train/test data
    length = len(runs[run][1].train) + len(runs[run][1].test) - 1
    
    # instantiate list for the forecasted dates
    new_dates = []
    
    # creates list of forecasted dates
    for h in range(1,steps+1):
        new_dates.append(max(runs[run][1].test.index)+QuarterEnd(h))
    
    # iterate through parameters for top 5 fits, fit the forecast dates
    for i, optimized in enumerate(runs[run][1].fit_parameters.values()):
        
        # predicting for each set of optimized parameters
        trial = optimized.predict(start = 0, end = length + steps)
        
        # resetting datetime index
        trial = trial[-steps:]
        trial.rename(dict(zip(trial[-steps:].index,new_dates)),inplace=True)
        
        # respective fit to dictionary
        predictions[i] = trial
    
    # prediction dict to dataframe
    forecast = pd.DataFrame.from_dict(predictions, orient='index').T
    
    # add column for mean of fits
    forecast['mean'] = forecast.mean(axis=1)
    
    # retain only average forecast per drug
    forecasts[run] = forecast['mean']

Finally, the summarized forecast data are collected and displayed.

In [79]:
# review forecasted data

# convert forecasts dict to dataframe
all_forecasts = pd.DataFrame.from_dict(forecasts, orient='index').T

all_forecasts['sum'] = all_forecasts.sum(axis=1)

# empty dict for SMAPE values
eval_dict = {}

# iterate through runs keys to get the SMAPEs form the Final_Results object
for run in runs.keys():
    eval_dict[run] = runs[run][1].test_eval

# dict of SMAPEs to df
eval_df = pd.DataFrame.from_dict(eval_dict, orient='index', columns = ['SMAPE'])

# concatenate the forecasts and SMAPEs
forecast_df = pd.concat([eval_df.T, all_forecasts], axis=0)

# display all data forecasted
display(forecast_df)

# display sums for the forecasts
print('Total forecasted prescriptions are shown below.')
display(all_forecasts['sum'])

# display summary of SMAPE data
display(eval_df.describe())

acetaminop    acetazolam        acetic     albuterol  \
SMAPE                1.387422e+01      0.567869      1.725540  1.643507e+00   
2022-12-31 00:00:00  1.478156e+06  71825.775893  14745.945715  7.229919e+06   
2023-03-31 00:00:00  1.586511e+06  71450.926578  14560.759659  7.319651e+06   
2023-06-30 00:00:00  1.699133e+06  70930.431095  14472.129435  7.280556e+06   
2023-09-30 00:00:00  1.817156e+06  70444.053043  14909.822396  7.277047e+06   
2022-09-30 00:00:00           NaN           NaN           NaN           NaN   

                       alclometas     aluminum    amantadine         amino  \
SMAPE                    3.596401    23.386176      0.805484      2.125834   
2022-12-31 00:00:00  10749.790247    91.074048  86581.589486  11475.511761   
2023-03-31 00:00:00  10708.372729  -470.582700  85391.827283  11437.219914   
2023-06-30 00:00:00  10585.600164 -1215.019281  85341.945290  11449.194947   
2023-09-30 00:00:00  10307.025874 -1875.438288  87049.421455  11411.463103   
2022-09-30 00:00:00           NaN          NaN           NaN           NaN   

                      aminocapro    amiodarone  ...     tobramycin  \
SMAPE                   2.152483      1.307709  ...       4.645990   
2022-12-31 00:00:00  2243.882867  91408.016570  ...  143530.810043   
2023-03-31 00:00:00  2263.514095  89693.281453  ...  145677.733463   
2023-06-30 00:00:00  2282.822739  88577.721064  ...  147474.225605   
2023-09-30 00:00:00  2301.869460  86757.095674  ...  148956.192919   
2022-09-30 00:00:00          NaN           NaN  ...            NaN   

                       triamcinol    triamteren   trimethobe   valproate  \
SMAPE                5.903138e-01      1.441616   112.062975    2.727497   
2022-12-31 00:00:00  2.210220e+06  75091.335719  -280.010874  470.294962   
2023-03-31 00:00:00  2.225385e+06  72198.759502 -1119.835887  454.490426   
2023-06-30 00:00:00  2.240525e+06  69276.400708 -1869.147477  456.786753   
2023-09-30 00:00:00  2.255655e+06  66743.474577 -2738.807992  455.689540   
2022-09-30 00:00:00           NaN           NaN          NaN         NaN   

                         verapamil  vibramycin        vitamin          zinc  \
SMAPE                     3.835583   10.353284       0.103236      1.820383   
2022-12-31 00:00:00  134830.508258  188.782911  686232.637542  26447.243602   
2023-03-31 00:00:00  131626.491652  -51.501162  685641.431074  26373.927848   
2023-06-30 00:00:00  128078.002431  192.321726  684967.164457  26247.923546   
2023-09-30 00:00:00  124751.640199  -60.500495  684209.895151  26209.875056   
2022-09-30 00:00:00            NaN         NaN            NaN           NaN   

                              sum  
SMAPE                         NaN  
2022-12-31 00:00:00  9.536447e+07  
2023-03-31 00:00:00  9.545443e+07  
2023-06-30 00:00:00  9.384251e+07  
2023-09-30 00:00:00  9.390578e+07  
2022-09-30 00:00:00  2.729141e+04  

[6 rows x 132 columns]

Total forecasted prescriptions are shown below.


2022-12-31    9.536447e+07
2023-03-31    9.545443e+07
2023-06-30    9.384251e+07
2023-09-30    9.390578e+07
2022-09-30    2.729141e+04
Name: sum, dtype: float64

SMAPE
count  131.000000
mean     6.391767
std     15.182006
min      0.029725
25%      1.039854
50%      2.297638
75%      4.785575
max    112.062975

In [110]:
91408.016570-86757.095674

4650.9208960000105

These data are discussed in the conclusions below. Select drugs are visualized below with their forecasted values.

In [102]:
select_drugs = ['acetaminop','albuterol','amantadine','amiodarone','doxycyclin']

for drug in select_drugs:
    train = runs[f'{drug}'][1].train
    test = runs[f'{drug}'][1].test
    total = pd.concat([train,test[:-1]], axis=0)
    forecasted = all_forecasts[f'{drug}']
    fig = go.Figure()
    fig.add_trace(go.Scatter(x=list(total.index), y = total[f'{drug}'], name = 'Original Data'))
    fig.add_trace(go.Scatter(x=list(forecasted.index), y = forecasted.values, name = 'Forecast Data'))
    fig.update_layout(title=f"{drug} Prescriptions Quarterly<Br>Real and Forecast")
    fig.update_xaxes(rangeslider_visible=True)
    fig.update_layout(xaxis_title = 'Year', yaxis_title='Prescriptions')
    fig.show()

In [107]:
# amiodarone figure for presentation

train = runs['amiodarone'][1].train
test = runs['amiodarone'][1].test
total = pd.concat([train,test[:-1]], axis=0)
forecasted = all_forecasts['amiodarone']
fig = go.Figure()
fig.add_trace(go.Scatter(x=list(total.index), y = total['amiodarone'], name = 'Original Data'))
fig.add_trace(go.Scatter(x=list(forecasted.index), y = forecasted.values, name = 'Forecast Data'))
fig.update_layout(title=f"Amiodarone Prescriptions Quarterly<Br>Real and Forecast")
fig.update_xaxes(rangeslider_visible=False)
fig.update_layout(xaxis_title = 'Year', yaxis_title='Prescriptions')
fig.show()

In [104]:
# amiodarone errors for presentation
print(runs['amiodarone'][1].train_eval)
print(runs['amiodarone'][1].test_eval)

11.464094290739336
1.3077087667120961


## Conclusions

131 drugs present in the Medicaid data and the shortage list have been forecast to the third quarter of 2023. The modeling will be improved in the future to be more robust in dealing with data where th drug is new to the Medicaid data set, and a full history is not available. A slight downward trend is expected in prescriptions overall, from 95.4 million in the final quarter of 2022 to 93.9 million in the third quarter of 2023. The SMAPE is shown for each prediction, and the mean SMAPE for all drugs is 6.4%. The lowest SMAPE is 0.03%, and 74% of forecasts have a SMAPE less than or equal to 4.8%. This suggests high validity of the forecasts.